### DISCUSSION TITLE
Using NavigationStateStore without a root view, i.e. path 1 is the root view

### DISCUSSION THREAD

**Discussion by bobspryn** [pointfreeco/swift-composable-architecture#2187] [Beta]
I'm building a wizard flow where any number of screens could be the first screen. It's easiest to think of these as a single collection of features, but the way the `NavigationStateStore` works means that I need to define a separate root view. I was trying to figure out a way to make it fully driven off the `StackState`. 

I mostly got it there, but this breaks the pop functionality:

<img width="900" alt="image" src="https://github.com/pointfreeco/swift-composable-architecture/assets/32696/37b58a88-41ba-46c8-a5f4-a43dd9915452">

Any thoughts on how we might do something like this?

**Comment by mbrandonw** [comment]
I mentioned this in Slack, but just to capture some of it here:

I think drawing inspiration from how UIKit works can help. With `UINavigationController` it only manages an array of controllers. There is no "root controller" like there is a root view in `NavigationStack`. Also the nav controller does extra work to make sure there is usually at least one controller in the `viewControllers` array.

For example, if `nav.viewControllers = [vc]`, and `vc` calls `popViewController`, then the nav controller just ignores that and keeps the view controller in the array. The only way for `viewControllers` to be empty is if you literally do `nav.viewControllers = []`.

So, that makes me think we can enforce an invariant in the root state that tries it's hardest to prevent the `path` from becoming empty:

```swift
struct AppFeature: Reducer {
  struct State {
    var path = StackState<Path>() {
      didSet {
        guard 
          self.path.isEmpty, 
          !oldValue.path.isEmpty 
        else { return }
        self.path = [oldValue.path.first!]
      }
    }
  }
  …
}
```

And then you would just use `EmptyView` inside the root view of `NavigationStack`.

It's of course not full proof. You can always create `AppFeature.State` with an empty `path`, but the same is true in UIKit too.

**Reply by Alex293** [reply]
I was going to suggest the empty view trick too but I thought this would cause issues with the menu that is displayed when you long press the back button. I haven’t tried thought.

**Reply by mbrandonw** [reply]
Yeah that's true. That will definitely be weird and I'm not sure much can be done about it. But, at least if someone does do that gesture, all they will see is "Back" at the bottom of the list, and tapping that will just take you back to the first screen, not all the way back to the empty view.

I also wonder how many people actually use that gesture ha 😬

**Reply by Alex293** [reply]
Yeah it’s your episodes that keep reminding me that it exists :)
